File: backend/core/updater.py
import os
import sys
import subprocess
import tempfile
import threading
from pathlib import Path
from typing import Optional, Callable
import requests

CURRENT_VERSION = "1.0.17"
GITHUB_REPO = "z0fans/NextDesk"
GITHUB_API_URL = f"https://api.github.com/repos/{GITHUB_REPO}/releases/latest"


class Updater:
    def __init__(self):
        self._download_progress: float = 0
        self._download_status: str = "idle"
        self._latest_version: Optional[str] = None
        self._download_url: Optional[str] = None
        self._download_thread: Optional[threading.Thread] = None

    def get_current_version(self) -> str:
        return CURRENT_VERSION

    def check_for_update(self) -> dict:
        try:
            resp = requests.get(GITHUB_API_URL, timeout=10)
            resp.raise_for_status()
            data = resp.json()

            tag = data.get("tag_name", "")
            latest = tag.lstrip("v")
            self._latest_version = latest

            assets = data.get("assets", [])
            exe_asset = next((a for a in assets if a["name"].endswith(".exe")), None)
            if exe_asset:
                self._download_url = exe_asset["browser_download_url"]

            has_update = self._compare_versions(latest, CURRENT_VERSION) > 0

            return {
                "has_update": has_update,
                "current_version": CURRENT_VERSION,
                "latest_version": latest,
                "download_url": self._download_url,
            }
        except Exception as e:
            return {
                "has_update": False,
                "current_version": CURRENT_VERSION,
                "latest_version": None,
                "error": str(e),
            }

    def get_download_status(self) -> dict:
        return {
            "status": self._download_status,
            "progress": self._download_progress,
        }

    def start_download(self) -> bool:
        if not self._download_url:
            return False
        if self._download_status == "downloading":
            return False

        self._download_status = "downloading"
        self._download_progress = 0
        self._download_thread = threading.Thread(
            target=self._download_update, daemon=True
        )
        self._download_thread.start()
        return True

    def _download_update(self):
        try:
            resp = requests.get(self._download_url, stream=True, timeout=300)
            resp.raise_for_status()

            total_size = int(resp.headers.get("content-length", 0))
            downloaded = 0

            temp_dir = tempfile.gettempdir()
            installer_path = Path(temp_dir) / "NextDesk_Update.exe"

            with open(installer_path, "wb") as f:
                for chunk in resp.iter_content(chunk_size=8192):
                    if chunk:
                        f.write(chunk)
                        downloaded += len(chunk)
                        if total_size > 0:
                            self._download_progress = (downloaded / total_size) * 100

            self._download_status = "ready"
            self._download_progress = 100

        except Exception as e:
            self._download_status = f"error: {str(e)}"
            self._download_progress = 0

    def install_update(self) -> bool:
        if self._download_status != "ready":
            return False

        temp_dir = tempfile.gettempdir()
        installer_path = Path(temp_dir) / "NextDesk_Update.exe"

        if not installer_path.exists():
            return False

        try:
            if sys.platform == "win32":
                subprocess.Popen(
                    [str(installer_path), "/SILENT"],
                    creationflags=subprocess.CREATE_NO_WINDOW,
                )
            return True
        except Exception:
            return False

    def _compare_versions(self, v1: str, v2: str) -> int:
        def parse(v):
            return [int(x) for x in v.split(".") if x.isdigit()]

        p1, p2 = parse(v1), parse(v2)
        for a, b in zip(p1, p2):
            if a > b:
                return 1
            if a < b:
                return -1
        return len(p1) - len(p2)

File: src-tauri/src/updater.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;

const CURRENT_VERSION: &str = env!("CARGO_PKG_VERSION");
const GITHUB_REPO: &str = "z0fans/NextDesk";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateInfo {
    pub has_update: bool,
    pub current_version: String,
    pub latest_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub download_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub fn get_current_version() -> String {
    CURRENT_VERSION.to_string()
}

pub async fn check_for_update() -> UpdateInfo {
    let url = format!(
        "https://api.github.com/repos/{GITHUB_REPO}/releases/latest"
    );
    let client = Client::builder()
        .timeout(Duration::from_secs(10))
        .build()
        .unwrap_or_default();

    let resp = match client
        .get(&url)
        .header("User-Agent", "NextDesk")
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => {
            return UpdateInfo {
                has_update: false,
                current_version: CURRENT_VERSION.into(),
                latest_version: None,
                download_url: None,
                error: Some(e.to_string()),
            };
        }
    };

    let data: Value = match resp.json().await {
        Ok(d) => d,
        Err(e) => {
            return UpdateInfo {
                has_update: false,
                current_version: CURRENT_VERSION.into(),
                latest_version: None,
                download_url: None,
                error: Some(e.to_string()),
            };
        }
    };

    let tag = data
        .get("tag_name")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let latest = tag.trim_start_matches('v').to_string();

    let download_url = data
        .get("assets")
        .and_then(|v| v.as_array())
        .and_then(|assets| {
            assets.iter().find_map(|a| {
                let name = a
                    .get("name")
                    .and_then(|n| n.as_str())
                    .unwrap_or("");
                if name.ends_with(".exe")
                    || name.ends_with(".dmg")
                {
                    a.get("browser_download_url")
                        .and_then(|u| u.as_str())
                        .map(|s| s.to_string())
                } else {
                    None
                }
            })
        });

    let has_update =
        compare_versions(&latest, CURRENT_VERSION) > 0;

    UpdateInfo {
        has_update,
        current_version: CURRENT_VERSION.into(),
        latest_version: Some(latest),
        download_url,
        error: None,
    }
}

fn compare_versions(v1: &str, v2: &str) -> i32 {
    let parse = |v: &str| -> Vec<i32> {
        v.split('.')
            .filter_map(|s| s.parse().ok())
            .collect()
    };
    let p1 = parse(v1);
    let p2 = parse(v2);
    for (a, b) in p1.iter().zip(p2.iter()) {
        if a > b {
            return 1;
        }
        if a < b {
            return -1;
        }
    }
    (p1.len() as i32) - (p2.len() as i32)
}
